session_test.go

```diff
@@ -0,0 +1,313 @@
+// +build windows
+
+package etw
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/Microsoft/go-winio/pkg/etw"
+	"github.com/stretchr/testify/suite"
+	"golang.org/x/sys/windows"
+)
+
+type testProvider struct {
+	suite.Suite
+
+	provider *etw.Provider
+	guid     windows.GUID
+}
+
+func (p *testProvider) SetupSuite() {
+	provider, err := etw.NewProvider("TestProvider", nil)
+	p.Require().NoError(err, "Failed to initialize test provider.")
+	p.provider = provider
+	p.guid = (windows.GUID)(provider.ID)
+}
+
+func (p *testProvider) TearDownSuite() {
+	p.Require().NoError(p.provider.Close(), "Failed to close test provider.")
+}
+
+func (p testProvider) generateEvents(ctx context.Context, levels []etw.Level, fields ...etw.FieldOpt) {
+	if fields == nil {
+		fields = etw.WithFields(etw.StringField("TestField", "Foo"))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			for _, l := range levels {
+				_ = p.provider.WriteEvent("TestEvent", etw.WithEventOpts(etw.WithLevel(l)), fields)
+			}
+		}
+	}
+}
+
+func (p *testProvider) TestSession() {
+	ctx, cancel := context.WithCancel(context.Background())
+	go p.generateEvents(ctx, []etw.Level{etw.LevelInfo})
+	defer func() {
+		cancel()
+	}()
+
+	session, err := NewSession(p.guid)
+	if err != nil {
+		p.Fail("Failed to create session; %s", err)
+	}
+
+	gotEvent := make(chan struct{})
+	var once sync.Once
+	cb := func(e *Event) {
+		once.Do(func() { close(gotEvent) })
+	}
+
+	done := make(chan struct{})
+	go func() {
+		err := session.Process(cb)
+		if err != nil {
+			p.Fail("Failed to start processing events; %s", err)
+		}
+		close(done)
+	}()
+
+	deadline := 10 * time.Second
+
+	// Wait for the first event from the provider.
+	select {
+	case <-gotEvent:
+		// Pass.
+	case <-time.After(deadline):
+		p.Fail("Failed to receive event from provider in 10s")
+	}
+
+	if err := session.Close(); err != nil {
+		p.Fail("Failed to close session properly; %s", err)
+	}
+
+	// Ensure processing goroutine wont block forever.
+	select {
+	case <-done:
+		// Pass.
+	case <-time.After(deadline):
+		p.Fail("Failed to stop event processing in 10s")
+	}
+}
+
+func (p *testProvider) TestUpdating() {
+	ctx, cancel := context.WithCancel(context.Background())
+	go p.generateEvents(ctx, []etw.Level{etw.LevelInfo, etw.LevelCritical})
+	defer func() {
+		cancel()
+	}()
+
+	session, err := NewSession(p.guid, WithLevel(TRACE_LEVEL_CRITICAL))
+	if err != nil {
+		p.Require().NoError(err, "Failed to create session")
+	}
+
+	gotCriticalEvent := make(chan struct{}, 1)
+	gotInformationEvent := make(chan struct{}, 1)
+	cb := func(e *Event) {
```

Comment:
Suppose some comments describing the testing scenario would be really helpful here